### PR TITLE
Add ref type to modal activator

### PR DESCRIPTION
**Pull request by dleroux**
### WHY are these changes introduced?

In order to improve accessibility on our Modal, this adds the ability to include a ref as the activator type on Modal.

We recently added the activator prop in order to return focus to it once the modal closes. However, it's not always possible to pass in the element and consumers are left to manage their own focus, which is in most cases overlooked. By adding a ref as a possible type, the Modal can still be responsible for focusing the correct node when closing.

### WHAT is this pull request doing?

- Adds React.RefObject as a possible prop type for the activator
- updates test
- updates documentation

## 

### How to 🎩

Storybook contains all the examples. Ensure they all work as expected.

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

**Review by alex-page — APPROVED** [review]

File: src/components/Modal/README.md [hunk 1]
```diff
@@ -641,11 +641,70 @@ function ModalWithScrollListenerExample() {
...
+    </div>
+  );
+}
+```
+
+### Modal without an activator prop
+
+<!-- example-for: web -->
+
+Use an external activator when technical limitations prevent you from passing the activator as element or a ref. Make sure to focus the activator on close when choosing this approach.
```
Comment (dleroux) [inline src/components/Modal/README.md:707]:
```suggestion
Use an external activator when technical limitations prevent you from passing the activator as an element or a ref. Make sure to focus the activator on close when choosing this approach.
```

File: src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -211,4 +219,10 @@ export const Modal: React.FunctionComponent<ModalProps> & {
   );
 };
 
+function isRef(
+  ref: React.RefObject<HTMLElement> | React.ReactElement,
+): ref is React.RefObject<HTMLElement> {
+  return Object.prototype.hasOwnProperty.call(ref, 'current');
```
Comment (kyledurand) [inline src/components/Modal/Modal.tsx:225]: Such a nice way to do a type guard 👍 

Just out of curiosity, does `ref.hasOwnProperty('current')` not work on react elements?

**Review by kyledurand — APPROVED** [review]
This looks great 👍

File: src/components/Modal/Modal.tsx (hunk 2 above)
Comment (dleroux) [reply src/components/Modal/Modal.tsx:225]: I think it would but we have this rule enabled: https://eslint.org/docs/rules/no-prototype-builtins

**Pull request merged**